server/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java

```diff
@@ -0,0 +1,252 @@
+  protected <T extends ExtractionNamespace> ListenableFuture<?> schedule(
+      final T namespace,
+      final ExtractionNamespaceFunctionFactory<T> factory,
+      final Runnable postRunnable,
+      final String cacheId
+  )
+  {
+    if (futures.containsKey(namespace.getNamespace())) {
+      // We also check at the end of the function, but fail fast here
+      throw new IAE("Namespace [%s] already exists! Leaving prior running", namespace.toString());
+    }
+    final Map<String, String> cache = getCacheMap(cacheId);
+    final Runnable runnable = factory.getCachePopulator(namespace, cache);
+
+    final ListenableFuture<?> future;
+    if (namespace.getUpdateMs() > 0) {
+      final long updateMs = namespace.getUpdateMs();
+      future = listeningScheduledExecutorService.scheduleAtFixedRate(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                if(!Thread.interrupted()) {
+                  runnable.run();
+                }
+                if(!Thread.interrupted()) {
+                  postRunnable.run();
+                }
+              }
+              catch (RuntimeException e) {
+                // Don't leave stale cache on error
+                delete(cacheId);
+                throw e;
+              }
+            }
+          },
+          updateMs - System.currentTimeMillis() % updateMs, updateMs, TimeUnit.MILLISECONDS
+      );
+    } else {
+      future = listeningScheduledExecutorService.schedule(runnable, 0, TimeUnit.MILLISECONDS);
+      Futures.addCallback(
+          future, new FutureCallback<Object>()
+          {
+            @Override
+            public void onSuccess(Object result)
+            {
+              try {
+                postRunnable.run();
+              }
+              catch (RuntimeException e) {
+                delete(cacheId);
+                throw e;
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              // Don't leave stale cache on error
+              delete(cacheId);
+              log.error(t, "Error populating namespace [%s]", namespace.toString());
+            }
+          }
+      );
+    }
+    final ListenableFuture<?> otherFuture = futures.putIfAbsent(namespace.getNamespace(), future);
+    if (otherFuture != null) {
+      future.cancel(true);
+      throw new IAE("Namespace [%s] already exists! Leaving prior running", namespace.toString());
+    }
+    return future;
+  }
+
+  protected abstract boolean swapAndClearCache(String namespaceKey, String cacheKey);
+
+  public abstract ConcurrentMap<String, String> getCacheMap(String namespaceOrCacheKey);
+
+  public boolean delete(final String ns)
+  {
+    if (!fnCache.containsKey(ns)) {
+      return false;
+    }
+    disabledNamespaces.add(ns);
+    final ListenableFuture<?> future = futures.get(ns);
+    future.cancel(false); // let it finish first.
```

Comment:
future.cancel(false) would return a boolean flag whether cancellation was successful or not. it might be good to do following.
```
if ( !future.cancel(false)) {
  throw some exception
}
```
and also do it before disabledNamespaces.add(ns)